content/blog/2023-12-11-lowering/index.md

```diff
@@ -0,0 +1,175 @@
+allocation. For each of these temps, this mapping class would assign it a unique offset in increments of 8 (starting from -8). 
+
+This was implemented as a dictionary in Python. So, if we come across a variable that is already in this dictionary, this means that we have seen it before and had already assigned it
+an offset, so we move on. Note that this happens on a per-function basis, so if we had two separate functions and they both contain a variable 'a', this variable is getting assigned 
+two separate offsets in each of these functions (for obvious reasons). By the end of this, for each function in our bril program, we have an object that has mapped each unique abstract variable to a unique stack location.
+
+The next step is to actually perform the trivial register allocation for each instruction in our function. As described above, this involves adding additional instructions before and after each instruction to load the appropriate temps off the stack, perform the operation, and then store the temps back onto the stack. The important property of when to add instructions before or after an instruction X is the following:
+
+- For each temporary variable that X reads from, we must include loads before the instruction to load these variables from their stack locations  
+- For the temporary variable that X writes to, we must include a store after the instruction to store this variable to its appropriate stack location
+
+These conditions led us to implement functions for each RISCV instruction object - one to return the list of abstract temps that the instruction reads from, and one to return the list of abstract temps that the instruction writes to. Knowing these for each instruction would allow us to determine where to place surrounding instructions to shuttle temps on and off the stack.
+
+Now, we iterate through each RISCV instruction and get the list of abstract temps that it reads from and writes to. For each of the abstract temps that the instruction reads from, we use our mapper to look through the dictionary and find the stack offsets for each. Before this instruction, we add loads from these stack offsets into our RISCV registers. Then we perform the operation in the instruction, being sure to replace the abstract temps that are read from with the RISCV registers that we just loaded into. Now that the abstract temps in the instruction that we read from are taken care of, we move on to writes. We look at the abstract temp we write to, and replace that in the instruction with the next available RISCV register we haven't used for the reads yet. We then find using the mapper the stack location of this destination variable and we add an instruction after the operation that performs a store from the destination RISCV register into the appropriate stack offset the mapper gave us.
```

Comment:
Embarrassingly enough, we simply didn't think of this...😅 This makes a ton of sense too since we know exactly which RISC-V instructions each Bril instruction maps to. This is definitely a good "work smarter not harder" moment for us to learn from lol.